taar_lite/recommenders/guid_based_recommender.py

```diff
@@ -17,22 +17,81 @@ class GuidBasedRecommender:
     passed as an input parameter based on periodically updated  addon-addon coinstallation
     frequency table generated from  Longitdudinal Telemetry data.
     This recommender will drive recommendations surfaced on addons.mozilla.org
+
+
+    We store the JSON data for the GUID coinstallation in memory. This
+    consumes ~ 15.8MB of heap.
+
+        In [10]: from pympler import asizeof
+
+        In [11]: jdata = json.load(open('guid_coinstallation.json'))
+
+        In [12]: asizeof.asizeof(jdata)
+        Out[12]: 15784672
+
+    Each of the data normalization dictionaries is also stored in
+    memory.
     """
+    _addons_coinstallations = None
+    _guid_maps = {}
+
     def __init__(self, ctx):
         self._ctx = ctx
         assert IS3Data in self._ctx
         self._init_from_ctx()
+        self._precompute_normalization()
 
     def _init_from_ctx(self):
         cache = self._ctx[IS3Data]
-        self.addons_coinstallations = cache.get_s3_json_content(ADDON_LIST_BUCKET,
-                                                                ADDON_LIST_KEY)
-        if self.addons_coinstallations is None:
+        self._addons_coinstallations = cache.get_s3_json_content(ADDON_LIST_BUCKET,
+                                                                 ADDON_LIST_KEY)
+        if self._addons_coinstallations is None:
             logger.error("Cannot download the addon coinstallation file {}".format(ADDON_LIST_KEY))
 
+    def _precompute_normalization(self):
+        if self._addons_coinstallations is None:
+            logger.error("Cannot find addon coinstallations to normalize.")
+            return
+
+        # Capture the total number of times that a guid was
+        # coinstalled with another GUID
+        #
+        # This is a map is guid->sum of coinstall counts
+        guid_count_map = {}
+
+        # Capture the number of times a GUID shows up per row
+        # of coinstallation data.
+        #
+        # This is a map of guid->rows that this guid appears on
+        guid_unique_coinstall = {}
+
+        guid_row_norm = {}
+
+        for guidkey, coinstalls in self._addons_coinstallations.items():
+            rowsum = sum(coinstalls.values())
+            for coinstall_guid, coinstall_count in coinstalls.items():
+
+                # Capture the total number of time a GUID was
+                # coinstalled with other guids
+                guid_count_map.setdefault(coinstall_guid, 0)
+                guid_count_map[coinstall_guid] += coinstall_count
+
+                # Capture the unique number of times a GUID is
+                # coinstalled with other guids
+                guid_unique_coinstall.setdefault(coinstall_guid, 0)
```

Comment:
Can you refer to `guid_unique_coinstall` as `row_count` to keep with the conventions we have used in discussing this method.

Reply:
yup. patched up now.